chain/lf3/manifest.go

```diff
@@ -84,9 +102,212 @@ func NewManifestProvider(mctx helpers.MetricsCtx, config *Config, cs *store.Chai
+			case <-t.C:
+				m, err := cmp.fetchActivationInfo(cmp.runningCtx)
+				if err != nil {
+					log.Warnw("got error while fetching manifest from contract", "error", err)
+					continue loop
+				}
+
+				if knownManifest.Equal(m) {
+					continue loop
+				}
+
+				if m != nil {
+					log.Infow("new manifest from contract", "enabled", true, "bootstrapEpoch", m.BootstrapEpoch)
+				} else {
+					log.Info("new manifest from contract", "enabled", false)
+				}
+				cmp.manifestChanges <- m
+				knownManifest = m
+			case <-cmp.runningCtx.Done():
+			}
+		}
+
+		return nil
+	})
+	return nil
+}
+
+func (cmp *ContractManifestProvider) fetchActivationInfo(ctx context.Context) (*manifest.Manifest, error) {
+	ethReturn, err := cmp.callContract(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("calling contract at %s: %w", cmp.address, err)
+	}
+	if len(ethReturn) == 0 {
+		return nil, nil
+	}
+
+	activationEpoch, compressedManifest, err := cmp.parseContractReturn(ethReturn)
+
+	if activationEpoch == math.MaxUint64 || len(compressedManifest) == 0 {
+		return nil, nil
+	}
+
+	reader := io.LimitReader(flate.NewReader(bytes.NewReader(compressedManifest)), 1<<20)
+	var m manifest.Manifest
+	err = json.NewDecoder(reader).Decode(&m)
+	if err != nil {
+		return nil, fmt.Errorf("got error while decoding manifest: %w", err)
+	}
+
+	if m.BootstrapEpoch < 0 || uint64(m.BootstrapEpoch) != activationEpoch {
+		return nil, fmt.Errorf("bootstrap epoch does not match: %d != %d", m.BootstrapEpoch, activationEpoch)
+	}
+
+	if err := m.Validate(); err != nil {
+		return nil, fmt.Errorf("manifest does not validate: %w", err)
+	}
+
+	if m.NetworkName != cmp.networkName {
+		return nil, fmt.Errorf("network name does not match, expected: %s, got: %s",
+			cmp.networkName, m.NetworkName)
+	}
+
+	return &m, nil
+}
+
+func (cmp *ContractManifestProvider) callContract(ctx context.Context) ([]byte, error) {
+	address, err := ethtypes.ParseEthAddress(cmp.address)
+	if err != nil {
+		return nil, fmt.Errorf("trying to parse contract address: %s: %w", cmp.address, err)
+	}
+
+	ethCall := ethtypes.EthCall{
+		To:   &address,
+		Data: must.One(ethtypes.DecodeHexString("0x2587660d")), // method ID of activationInformation()
+	}
+
+	fMessage, err := ethCall.ToFilecoinMessage()
+	if err != nil {
+		return nil, fmt.Errorf("converting to filecoin message: %w", err)
+	}
+
+	msgRes, err := cmp.stateCaller.StateCall(ctx, fMessage, types.EmptyTSK)
+	if err != nil {
+		return nil, fmt.Errorf("state call error: %w", err)
+	}
+	if msgRes.MsgRct.ExitCode != 0 {
+		return nil, fmt.Errorf("message returned exit code %v: %v", msgRes.MsgRct.ExitCode, msgRes.Error)
+	}
+
+	var ethReturn abi.CborBytes
+	err = ethReturn.UnmarshalCBOR(bytes.NewReader(msgRes.MsgRct.Return))
+	if err != nil {
+		return nil, fmt.Errorf("could not decode return value: %w", err)
+	}
+	return []byte(ethReturn), nil
+}
+
+func (cmp *ContractManifestProvider) parseContractReturn(retBytes []byte) (uint64, []byte, error) {
```

Comment:
We want tests for these please. Hard coded binary that gets parsed just to check for sanity. Otherwise, reviewing this code is hard.

Reply:
Done